brvs-rules/src/main/java/iroha/validation/rules/impl/billing/BillingRule.java

```diff
@@ -0,0 +1,251 @@
+  private static final BigDecimal NO_FEE = BigDecimal.ZERO;
+
+  private boolean isRunning = false;
+  private final String getBillingURL;
+  private final String rmqHost;
+  private final int rmqPort;
+  private final String rmqExchange;
+  private final String rmqRoutingKey;
+  private final Gson gson = new Gson();
+  private final Set<BillingInfo> cache = new HashSet<>();
+
+  public BillingRule(String getBillingURL,
+      String rmqHost,
+      int rmqPort,
+      String rmqExchange,
+      String rmqRoutingKey) throws IOException {
+
+    if (Strings.isNullOrEmpty(getBillingURL)) {
+      throw new IllegalArgumentException("Billing URL must not be neither null nor empty");
+    }
+    if (Strings.isNullOrEmpty(rmqHost)) {
+      throw new IllegalArgumentException("RMQ host must not be neither null nor empty");
+    }
+    if (rmqPort < 1 || rmqPort > 65535) {
+      throw new IllegalArgumentException("RMQ port must be valid");
+    }
+    if (Strings.isNullOrEmpty(rmqExchange)) {
+      throw new IllegalArgumentException("RMQ exchange must not be neither null nor empty");
+    }
+    if (Strings.isNullOrEmpty(rmqRoutingKey)) {
+      throw new IllegalArgumentException("RMQ routing key must not be neither null nor empty");
+    }
+
+    this.getBillingURL = getBillingURL;
+    this.rmqHost = rmqHost;
+    this.rmqPort = rmqPort;
+    this.rmqExchange = rmqExchange;
+    this.rmqRoutingKey = rmqRoutingKey;
+    runCacheUpdater();
+  }
+
+  private void runCacheUpdater() throws IOException {
+    if (isRunning) {
+      return;
+    }
+    isRunning = true;
+    readBillingOnStartup();
+    getMqUpdatesObservable().subscribeOn(Schedulers.from(Executors.newSingleThreadExecutor()))
+        .subscribe(update -> {
+              final BillingInfo currentBillingInfo = cache
+                  .stream()
+                  .filter(entry -> entry.equals(update))
+                  .findAny()
+                  .orElse(null);
+              if (currentBillingInfo == null
+                  || currentBillingInfo.getUpdated().isBefore(update.getUpdated())) {
+                // Will be replaced, see BillingInfo#hashCode
+                cache.add(update);
+              }
+            }
+        );
+  }
+
+  private void readBillingOnStartup() throws IOException {
+    final JsonObject root = jsonParser.parse(executeGetRequest()).getAsJsonObject();
+    for (BillingTypeEnum billingType : BillingTypeEnum.values()) {
+      final String label = billingType.label;
+      cache.addAll(
+          BillingInfo.parseBillingHttpDto(
+              label,
+              gson.fromJson(
+                  root.getAsJsonObject(label),
+                  new TypeToken<HashMap<String, HashMap<String, JsonObject>>>() {
+                  }.getType()
+              )
+          )
+      );
+    }
+  }
+
+  private String executeGetRequest() throws IOException {
+    URL url = new URL(getBillingURL);
```

Comment:
Maybe use some framework for http requests for simplicity? Like feign.

Reply:
Relevant point, but I'm really not sure if I should do it since `brvs-rules` is really intended to use as less dependencies as possible. I think we should add it if there are more HTTP related code, but for now it is about several lines of code only.